schemaregistry/schemaregistry.md

```diff
@@ -0,0 +1,257 @@
+# CNCF Schema Registry API Version 0.1-rc01s
+
+## Abstract
+
+This specification defines a simple API for storing, organizing, and accessing
+schema documents for use with serialization and data validation frameworks such
+as Apache Avro and JSON Schema.
+
+## Status of this document
+
+This document is a working draft.
+
+## 1. Introduction
+
+Schema-dependent serialization formats and related frameworks such as Apache
+Avro are popular in eventing and messaging scenarios. With most of the metadata
+information such as names and types of fields being held externally in schema
+documents, these frameworks are able to produce very compact on-wire
+representations of structured data that is carries as the payload of events and
+messages. While the wire-footprint savings are often very significant, the added
+cost is that of schema handling. All communicating parties who need to
+deserialize the event or message payload need access to the schema used to
+publish it.
+
+For formats that do not depend on external schemas for serialization, schemas
+might still be required in some scenarios to allow for consumers or inspecting
+intermediaries to validate the data structure's compliance with a set of rules.
+
+This specification defines a very simple API focused on storing, organizing, and
+accessing such documents. While there are very sophisticated metadata management
+solutions available as open source und under neutral governance already, the focus
+of this specification is not great sophistication but straightforward simplicity.
+
+This schema registry deals with only three top-level elements:
+
+- Schema Group: A named collection of schemas. Each group holds a logically
+  related set of schemas, typically belonging to a particular application or
+  subject to shared access control rules.
+- Schema: A document describing the structure, names, and types of some
+  structured data payload.
+- Version: A specific version of a schema document. Even though not prescribed
+  in this specification, an implementation might choose to impose compatibility
+  constraints on versions following the initial version of a schema.
+
+> NOTE: There are two fundamentally different options for how we can organize
+> multiple schema formats and this specification draft reflects both: 
+>
+> 1) In an "ideal" REST service, any data entity can be represented ("R" in
+>    REST) in multiple different data formats, which a client can request in
+>    prioritized order during content-type negotiation. If a service offers up
```

Comment:
This also would imply equal expressiveness of all supported schema languages, which is not necessarily the case (type systems might be different etc.) Also from what I see from Debezium usage (used a lot with Kafka schema registries like Red Hat's and Confluent's), people seem fine with one format or another for a given event type.

Reply:
@gunnarmorling I'm also generally leaning towards the simpler option, but I don't think schema languages needs to be perfectly aligned into all corners of their features for this to make sense. The majority of data structurs being passed around are fairly straightforward and do ok with a handful of common types and without getting much into sophisticated constraining expressions. Having one transfer-optimized format for a particular data structure (Avro, Parquet, Proto, etc) and then also a JSON representation for broader interop reach isn't terrible.